pymc/progress_bar.py

```diff
@@ -192,6 +192,143 @@ def callbacks(self, task: "Task"):
+    speed = draws / max(elapsed, 1e-6)
+
+    if speed > 1 or speed == 0:
+        unit = "draws/s"
+    else:
+        unit = "s/draws"
+        speed = 1 / speed
+
+    return speed, unit
+
+
+def create_rich_progress_bar(full_stats, step_columns, progressbar, progressbar_theme):
+    columns = [TextColumn("{task.fields[draws]}", table_column=Column("Draws", ratio=1))]
+
+    if full_stats:
+        columns += step_columns
+
+    columns += [
+        TextColumn(
+            "{task.fields[sampling_speed]:0.2f} {task.fields[speed_unit]}",
+            table_column=Column("Sampling Speed", ratio=1),
+        ),
+        TimeElapsedColumn(table_column=Column("Elapsed", ratio=1)),
+        TimeRemainingColumn(table_column=Column("Remaining", ratio=1)),
+    ]
+
+    return CustomProgress(
+        RecolorOnFailureBarColumn(
+            table_column=Column("Progress", ratio=2),
+            failing_color="tab:red",
+            complete_style=Style.parse("rgb(31,119,180)"),  # tab:blue
+            finished_style=Style.parse("rgb(31,119,180)"),  # tab:blue
+        ),
+        *columns,
+        console=Console(theme=progressbar_theme),
+        disable=not progressbar,
+        include_headers=True,
+    )
+
+
+class MarimoProgressTask:
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+    @property
+    def chain_idx(self) -> int:
+        return self.kwargs.get("chain_idx", 0)
+
+    @property
+    def total(self):
+        return self.kwargs.get("total", 0)
+
+    @property
+    def elapsed(self):
+        return self.kwargs.get("elapsed", 0)
+
+
+class MarimoProgressBar:
+    def __init__(self) -> None:
+        self.tasks: list[ProgressTask] = []
+        self.divergences: dict[int, int] = {}
+
+    def __enter__(self):
+        """Enter the context manager."""
+        import marimo as mo
+
+        total_draws = (self.tasks[0].total + 1) * len(self.tasks)
+
+        self.bar = mo.status.progress_bar(total=total_draws, title="Sampling PyMC model")
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the context manager."""
+        self.bar._finish()
+
+    def add_task(self, *args, **kwargs):
+        """Add a task to the progress bar."""
+        task = MarimoProgressTask(*args, **kwargs)
+        self.tasks.append(task)
+        return task
+
+    def update(self, task_id, **kwargs):
+        """Update the task with the given ID with the provided keyword arguments."""
+        if self.bar.progress.current >= cast(int, self.bar.progress.total):
+            return
+
+        self.divergences[task_id.chain_idx] = kwargs.get("divergences", 0)
+
+        total_divergences = sum(self.divergences.values())
+
+        update_kwargs = {}
+        if total_divergences > 0:
+            word = "draws" if total_divergences > 1 else "draw"
+            update_kwargs["subtitle"] = f"{total_divergences} diverging {word}"
+
+        self.bar.progress.update(**update_kwargs)
```

Comment:
We changed the logic to not be obsessed with NUTS. There is now a "failing variable" that the step methods report back, and that's what the ProgressBar manager can then use to indicate sampling is failing.

This reverts back to over-specializing for nuts